ir/function.cpp

```diff
@@ -428,5 +438,115 @@ void DomTree::printDot(std::ostream &os) const {
   os << "}\n";
 }
 
+
+void LoopAnalysis::getDepthFirstSpanningTree() {
+  std::map<const BasicBlock *, std::vector<const BasicBlock *>> edges;
+  unsigned current = 0;
+  for (const auto &[src, dst, instr] : cfg) {
+    edges[&src].push_back(&dst);
+  }
+
+  std::function<void(const BasicBlock *)> dfs = [&](const BasicBlock *bb) {
+    node.push_back(bb);
+    number[bb] = current++;
+    for (auto tgt : edges [bb]) {
+      if (number.find(tgt) == number.end()) {
+        dfs(tgt);
+      }
+    }
+    last[number[bb]] = current - 1;
+  };
+
+  dfs(&f.getFirstBB());
+}
+
+void LoopAnalysis::printNodeType(ostream &os, const NodeType &t) const {
+  switch (t) {
+  case nonheader: os<<"n."; break;
+  case self:      os<<"s."; break;
+  case reducible: os<<"r."; break;
+  }
+}
+
+// Tarjan, R. (1974).
+// Testing Flow Graph Reducibility.
+void LoopAnalysis::analysis() {
+  getDepthFirstSpanningTree();
+  unsigned bb_count = f.getBBs().size();
+
+  auto isAncestor = [this](unsigned w, unsigned v) -> bool {
+    return w <= v && v <= last.at(w);
+  };
+
+  vector<set<unsigned>> nonBackPreds, backPreds;
+
+  for (unsigned w = 0; w < bb_count; ++w) {
+    nonBackPreds.push_back({});
+    backPreds.push_back({});
+    header.push_back(0);
+    type.push_back(NodeType::nonheader);
+    for (const auto &[src, dst, instr] : cfg) {
+      (void) instr;
+      if (&dst == node[w]) {
+        unsigned v = number.at(&src);
+        if (isAncestor(w, v))
+          backPreds[w].insert(v);
+        else
+          nonBackPreds[w].insert(v);
+      }
+    }
+  }
+  UnionFind uf(bb_count);
+
+  for (unsigned w = bb_count - 1; w != static_cast<unsigned>(-1); --w) {
+    set<unsigned> P;
+    for (unsigned v : backPreds[w])
+      if (v != w)
+        P.insert(uf.find(v));
+      else
+        type[w] = NodeType::self;
+
+    if (!P.empty()) type[w] = NodeType::reducible;
+    set<unsigned> workList(P);
+
+    while (!workList.empty()) {
+      unsigned x = *workList.begin();
+      workList.erase(x);
+
+      for (unsigned y : nonBackPreds[x]) {
+        unsigned yy = uf.find(y);
+        if (!isAncestor(w, yy)) {
+          UNREACHABLE();
```

Comment:
Nuno: what's the best way for terminating with an error message in Alive2, is assert(0 && "irreducible loop detected"); a good choice here?

Reply:
the code shouldn't crash with irreducible loops. It can produce a sub-optimal result, but shouldn't crash.

Reply:
also, never use assert for error handling (although I believe we leave asserts enabled even in release builds). if you were going to exit here, you could do something like 
```   
 if (has_failure) {
      cerr << "Alive2: Transform doesn't verify; aborting!" << endl;
      exit(1);
    }
```

Reply:
@regehr correct!
@nunoplopes Need to think if it is safe to remove this line, or in another word, if continue going without break could give wrong result.

Reply:
maybe write the 3 missing lines from Havlak and call it a day?